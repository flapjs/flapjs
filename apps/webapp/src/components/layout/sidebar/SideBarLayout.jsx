import React from 'react';
import Style from './SideBarLayout.module.css';

import * as SideHelper from './SideHelper.js';

/**
 * 
 * @param {object} props 
 * @param {string} [props.className]
 * @param {object} [props.style]
 * @param {React.ReactNode} [props.children]
 * @param {React.FC} props.sideBar
 * @param {'left'|'right'|'top'|'bottom'} [props.side]
 * @param {number|string} [props.sideMargin]
 * @param {'hidden'|'auto'} [props.viewOverflow]
 */
export function SideBarLayout(props) {
    const {
        className = '',
        style = {},
        side = 'left',
        sideBar,
        sideMargin = 'auto',
        viewOverflow = 'hidden'
    } = props;

    const containerFlexDirection = (
        side === 'left'
            ? 'row'
            : side === 'right'
                ? 'row-reverse'
                : side === 'top'
                    ? 'column'
                    : 'column-reverse'
    );

    const contentFlexDirection = (
        SideHelper.isHorizontal(side)
            ? 'column'
            : 'row'
    );

    const cssVars = {
        '--side-margin': sideMargin,
    };

    return (
        <div className={`${Style.container} ${className || ''}`}
            style={{ flexDirection: containerFlexDirection, ...cssVars, ...style }}>
            <div className={`${Style.sideBar} ${side}`}>
                <div className={`${Style.sideContent}`}
                    style={{ flexDirection: contentFlexDirection }}>

                    {renderSideBarContent(sideBar, side)}

                </div>
            </div>
            <div className={Style.viewport}
                style={{ overflow: viewOverflow }}>
                {props.children}
            </div>
        </div>
    );
}

function renderSideBarContent(sideBar, side) {
    if (typeof sideBar === 'function') {
        return sideBar(side);
    }
    else {
        return (
            <div>
                {sideBar}
            </div>
        );
    }
}
