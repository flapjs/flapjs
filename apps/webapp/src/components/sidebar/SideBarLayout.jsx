import React from 'react';
import PropTypes from 'prop-types';
import Style from './SideBarLayout.module.css';

import * as SideHelper from './SideHelper.js';

export default function SideBarLayout(props)
{
    const { className, style, side, sideBar, sideMargin, viewOverflow } = props;

    const containerFlexDirection = (
        side === 'left'
            ? 'row'
            : side === 'right'
                ? 'row-reverse'
                : side === 'top'
                    ? 'column'
                    : 'column-reverse'
    );

    const contentFlexDirection = (
        SideHelper.isHorizontal(side)
            ? 'column'
            : 'row'
    );

    const cssVars = {
        '--side-margin': sideMargin,
    };

    return (
        <div className={`${Style.container} ${className || ''}`}
            style={{ flexDirection: containerFlexDirection, ...cssVars, ...style }}>

            <div className={`${Style.sideBar} ${side}`}>
                <div className={`${Style.sideContent}`}
                    style={{ flexDirection: contentFlexDirection }}>
                    
                    {renderSideBarContent(sideBar, side)}

                </div>
            </div>

            <div className={Style.viewport}
                style={{ overflow: viewOverflow }}>
                {props.children}
            </div>
            
        </div>
    );
}
SideBarLayout.propTypes = {
    className: PropTypes.string,
    style: PropTypes.object,
    children: PropTypes.node,
    sideBar: PropTypes.func,
    side: PropTypes.oneOf(['left', 'right', 'top', 'bottom']),
    sideMargin: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
    ]),
    viewOverflow: PropTypes.oneOf(['hidden', 'auto']),
};
SideBarLayout.defaultProps = {
    side: 'left',
    sideMargin: 'auto',
    viewOverflow: 'hidden',
};

function renderSideBarContent(sideBar, side)
{
    if (typeof sideBar === 'function')
    {
        return sideBar(side);
    }
    else
    {
        return (
            <div>
                {sideBar}
            </div>
        );
    }
}
