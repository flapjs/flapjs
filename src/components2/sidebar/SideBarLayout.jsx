import React from 'react';
import PropTypes from 'prop-types';
import Style from './SideBarLayout.module.css';

/**
 * A React component that adds a sidebar with a viewport.
 */
class SideBarLayout extends React.Component
{
    constructor(props)
    {
        super(props);
    }

    /** @override */
    render()
    {
        const props = this.props;

        const side = props.side;
        
        const horizontal = side === 'left' || side === 'right';

        const containerFlexDirection = (
            side === 'left' ? 'row' :
                side === 'right' ? 'row-reverse' :
                    side === 'top' ? 'column' :
                        'column-reverse'
        );
        
        return (
            <div className={Style.container + ' ' + (props.className || '')} style={{ flexDirection: containerFlexDirection }}>
                <div className={Style.sidebar + ' ' + side}>
                    {props.renderSideBar && props.renderSideBar(this, horizontal)}
                </div>
                <div className={Style.viewport}>
                    {props.children}
                </div>
            </div>
        );
    }
}

SideBarLayout.propTypes = {
    className: PropTypes.string,
    children: PropTypes.node,
    renderSideBar: PropTypes.func,
    side: PropTypes.oneOf([
        'left',
        'right',
        'top',
        'bottom',
    ]),
};
SideBarLayout.defaultProps = {
    side: 'left',
};

export default SideBarLayout;
