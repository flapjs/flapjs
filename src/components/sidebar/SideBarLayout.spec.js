import React from 'react';
import { shallow } from 'enzyme';

import SideBarLayout from './SideBarLayout.jsx';

let wrapper;

beforeEach(() =>
{
    wrapper = shallow(<SideBarLayout/>);
});

describe('<SideBarLayout/> rendering', () =>
{
    test('renders correctly', () =>
    {
        expect(wrapper).toMatchSnapshot();
    });
});

describe('<SideBarLayout/> interaction', () =>
{
    /** I promise I will test you soon... */
});

describe('<SideBarLayout/> lifecycle', () =>
{
    /** ...later. */
});
