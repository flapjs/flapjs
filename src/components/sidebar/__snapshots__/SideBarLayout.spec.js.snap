// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideBarLayout/> rendering renders correctly 1`] = `
<div
  className="container "
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <div
    className="sidebar left"
  />
  <div
    className="viewport"
  />
</div>
`;
